lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs

```diff
@@ -0,0 +1,1015 @@
+        /// names are in ASCII 8.
+        const CB_ASCII_NAMES = 0x0004;
+    }
+}
+
+/// This header (CLIPRDR_HEADER) is present in all clipboard PDUs.
+#[derive(Debug, PartialEq, Eq)]
+struct ClipboardPDUHeader {
+    /// Specifies the type of clipboard PDU that follows the dataLen field.
+    msg_type: ClipboardPDUType,
+    msg_flags: ClipboardHeaderFlags,
+    /// Specifies the size, in bytes, of the data which follows this header.
+    data_len: u32,
+}
+
+impl ClipboardPDUHeader {
+    fn new(msg_type: ClipboardPDUType, msg_flags: ClipboardHeaderFlags, data_len: u32) -> Self {
+        ClipboardPDUHeader {
+            msg_type,
+            msg_flags,
+            data_len,
+        }
+    }
+
+    fn decode(payload: &mut Payload) -> RdpResult<Self> {
+        let typ = payload.read_u16::<LittleEndian>()?;
+        Ok(Self {
+            msg_type: ClipboardPDUType::from_u16(typ)
+                .ok_or_else(|| invalid_data_error(&format!("invalid message type {:#04x}", typ)))?,
+            msg_flags: ClipboardHeaderFlags::from_bits(payload.read_u16::<LittleEndian>()?)
+                .ok_or_else(|| invalid_data_error("invalid flags in clipboard header"))?,
+            data_len: payload.read_u32::<LittleEndian>()?,
+        })
+    }
+
+    fn encode(&self) -> RdpResult<Vec<u8>> {
+        let mut w = vec![];
+        w.write_u16::<LittleEndian>(self.msg_type as u16)?;
+        w.write_u16::<LittleEndian>(self.msg_flags.bits())?;
+        w.write_u32::<LittleEndian>(self.data_len)?;
+        Ok(w)
+    }
+}
+#[derive(Clone, Copy, Debug, Eq, PartialEq, FromPrimitive, ToPrimitive)]
+#[allow(non_camel_case_types)]
+enum ClipboardPDUType {
+    CB_MONITOR_READY = 0x0001,
+    CB_FORMAT_LIST = 0x0002,
+    CB_FORMAT_LIST_RESPONSE = 0x0003,
+    CB_FORMAT_DATA_REQUEST = 0x0004,
+    CB_FORMAT_DATA_RESPONSE = 0x0005,
+    CB_TEMP_DIRECTORY = 0x0006,
+    CB_CLIP_CAPS = 0x0007,
+    CB_FILECONTENTS_REQUEST = 0x0008,
+    CB_FILECONTENTS_RESPONSE = 0x0009,
+    CB_LOCK_CLIPDATA = 0x000A,
+    CB_UNLOCK_CLIPDATA = 0x000B,
+}
+
+/// An optional PDU (CLIPRDR_CAPS) used to exchange capability information.
+/// If this PDU is not sent, it is assumed that the endpoint which did not
+/// send capabilities is using the default values for each field.
+#[derive(Debug)]
+struct ClipboardCapabilitiesPDU {
+    // The protocol is written in such a way that there can be
+    // a variable number of capability sets in this PDU. However,
+    // the spec only defines one type of capability set (general),
+    // so we'll just use an Option.
+    general: Option<GeneralClipboardCapabilitySet>,
+}
+
+const CB_CAPS_VERSION_2: u32 = 0x0002;
+
+impl ClipboardCapabilitiesPDU {
+    fn encode(&self) -> RdpResult<Vec<u8>> {
+        let mut w = vec![];
+        // there's either 0 or 1 capability sets included here
+        w.write_u16::<LittleEndian>(self.general.as_ref().map_or(0, |_| 1))?;
```

Comment:
This looks very confusing at glance.

```suggestion
        w.write_u16::<LittleEndian>(self.general.is_some() as u16)?;
```

Reply:
You can convert a bool to u16 like that? TIL